General optimizations

itzexor opened the pull request:
Simple profiling with the GJS/CJS profiler revealed:
1. paths being called relatively frequently when the features they support weren't enabled:
-magnifier cursor update
-desklet mouse tracker
2. window restack connections are always a top consumer on general non-cinnamon-ui usage
3. <s>uiGroup is always allocating, and the allocate does nothing beyond default</s>
4. notification daemon is always checking the WindowTracker to clear notifications, even when there are none

itzexor commented on js/ui/main.js:
```diff
@@ -350,23 +350,8 @@ function start() {
-                        [alloc.min_size, alloc.natural_size] = [height, height];
-                    });
+    uiGroup = new Clutter.Actor({ name: 'uiGroup' });
+    global.stage.connect('notify::size', () => { uiGroup.size = global.stage.size });
```
fixme: replace with bind constraint like the one used for stage_bg actor below

jaszhix commented:
Just for reference, this was closed after discussions on Slack. I did not intend for this to be closed. There were two problems I pointed out:

- 885582d switches uiGroup to an StWidget actor, but this loses the skip_paint optimization in CinnamonGenericContainer, and seemed less responsive. I tried to [benchmark](https://openbenchmarking.org/result/1902033-SK-VARIOUSJA09&amp;obr_sor=y) this with gputest, which might at least show the impact Cinnamon has on the GPU.
- 2dd1532 slows the cursor movement. I'm not sure why. It seems like a muffin or xorg bug which had less impact while Cinnamon was polling XQueryPointer.

itzexor commented:
- rebased on master
- dropped uiGroup commit to another branch for future investigation
- fixed up magnifier commit to cover hopefully all cases where it can be first called.

i think the desklet one definitely wastes resources if you're not using desklets. i'd like to keep that one in if possible.

jaszhix commented on js/ui/deskletManager.js:
```diff
@@ -252,7 +253,7 @@ function _onEnabledDeskletsChanged() {
     // Make sure all desklet extensions are loaded.
     // Once loaded, the desklets will add themselves via finishExtensionLoad
-    initEnabledDesklets();
+    initEnabledDesklets().then(updateMouseTracking);
```
The definitions array handling is synchronous and is already determined before we start loading desklets, so we don't need to call this through the promise prototype. The promise is only needed if we want to have something run after every desklet is loaded from disk. The cursor movement improves when I call updateMouseTracking before initEnabledDesklets. Maybe a CJS/GLib bug (since the promise itself is implemented as an idle source)?

itzexor replied:
I have no idea why it would have that effect, so I can't really say

jaszhix commented on js/ui/deskletManager.js:
```diff
@@ -77,7 +77,7 @@ function init(){
 
         deskletsLoaded = true;
-        enableMouseTracking(true);
+        updateMouseTracking();
```
This could be moved to line 73 after `definitions = getDefinitions()`. The array won't change after that point until a desklet is added or removed.

itzexor replied:
The mouse tracker source calls methods on the desklet instances, so I figured after the promise was the proper place to put it to ensure the function could only ever be called after the first instance was fully loaded. Granted it would have to take > 500ms for the instance to load for it to cause an issue, is my logic incorrect here? I don't really know the internals of the extensions system.

OTOH if definitions[i].desklet is falsy before it fully loads, calling it immediately should be ok.

jaszhix replied:
I didn't see that it was checking the desklet itself, that makes sense to me.

itzexor commented:
- rebased on master
- fixed mistake in magnifier.js where I replaced `new Object(roi)` with `Object.Clone(roi)` (non existent function) instead of `Object.assign({}, roi)` as I meant to. This gets rid of a "use an object literal" warning.
